File: contracts/src/main/java/ch/maxant/kdc/contracts/Contract.java
package ch.maxant.kdc.contracts;

import ch.maxant.kdc.products.WithValidity;
import ch.maxant.kdc.products.Product;
import org.hibernate.annotations.Type;
import org.hibernate.envers.Audited;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.UUID;

@Audited
@Entity
@Table(name = "CONTRACTS")
public class Contract implements WithValidity {

    @Id
    @Column(name = "ID")
    @Type(type="uuid-char")
    private UUID id = UUID.randomUUID();

    @Column(updatable = false, nullable = false, name = "CONTRACTNUMBER")
    private String contractNumber;

    @Version
    @Column(nullable = false, name = "VERSION")
    private Integer version;

    @Column(nullable = false, name = "FROM_")
    private LocalDateTime from;

    @Column(nullable = false, name = "TO_")
    private LocalDateTime to;

    /** an optional version of the product instance, matching the date for which the contract version and product instance was loaded */
    @Transient
    private Product product;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getContractNumber() {
        return contractNumber;
    }

    public void setContractNumber(String contractNumber) {
        this.contractNumber = contractNumber;
    }

    public Integer getVersion() {
        return version;
    }

    public LocalDateTime getFrom() {
        return from;
    }

    public void setFrom(LocalDateTime from) {
        this.from = from;
    }

    public LocalDateTime getTo() {
        return to;
    }

    public void setTo(LocalDateTime to) {
        this.to = to;
    }

    public Product getProduct() {
        return product;
    }

    public void setProduct(Product product) {
        this.product = product;
    }
}

File: mf/contracts/src/main/kotlin/ch/maxant/kdc/mf/contracts/entity/Contract.kt
package ch.maxant.kdc.mf.contracts.entity

import org.hibernate.annotations.Type
import java.time.LocalDateTime
import java.util.*
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.Table

@Entity
@Table(name = "T_CONTRACTS")
open class Contract( // add open, rather than rely on maven plugin, because @QuarkusTest running in IntelliJ seems to think its final

    @Id
    @Column(name = "ID")
    @Type(type = "uuid-char")
    open var id: UUID,

    @Column(name = "PRODUCT_ID")
    open var productId: String,

    @Column(name = "STARTTIME")
    open var start: LocalDateTime,

    @Column(name = "ENDTIME")
    open var end: LocalDateTime

) {
    constructor(): this(UUID.randomUUID(), "", LocalDateTime.MIN, LocalDateTime.MAX)
}